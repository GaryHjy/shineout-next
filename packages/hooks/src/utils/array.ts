export const flattenArray = <T>(arr1: any[]): T[] =>
  arr1.reduce(
    (acc, val) => (Array.isArray(val) ? acc.concat(flattenArray(val)) : acc.concat(val)),
    [],
  );
