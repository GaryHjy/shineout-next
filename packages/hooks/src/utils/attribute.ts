export const getDataAttribute = (attrs: Record<string, string | undefined>) => {
  return Object.keys(attrs).reduce((acc, key) => {
    const ns = `data-soui-${key}`;
    if (attrs[key] !== undefined) {
      acc[ns] = attrs[key]!;
    }
    return acc;
  }, {} as Record<string, string>);
};

export const getDataAttributeName = (name: string) => {
  return `data-soui-${name}`;
};
