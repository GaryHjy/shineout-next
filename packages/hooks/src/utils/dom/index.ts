export { default, default as NormalizeWheel } from './normalize-wheel';
export { parsePxToNumber, addResizeObserver } from './element';
export type { OldWheelEvent } from './normalize-wheel.type';
