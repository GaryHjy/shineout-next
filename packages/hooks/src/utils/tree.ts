export const getFlattenTree = <T>(
  data: T[],
  childrenKey = 'children' as keyof T,
  wide?: boolean,
) => {
  const arr: T[][] = [];
  const flatten = (list: T[], path: T[]) => {
    list.forEach((item) => {
      const children = item[childrenKey] as any;
      if (children && children.length > 0) {
        const clonedPath = [...path];
        clonedPath.push(item);
        if (wide) arr.push(clonedPath);
        flatten(children, clonedPath);
      } else {
        arr.push([...path, item]);
      }
    });
  };
  flatten(data, []);
  return arr;
};
