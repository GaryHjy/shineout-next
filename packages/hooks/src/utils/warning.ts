const deprecated = (
  prop: string,
  newProp: string,
  component: string,
  extraMessage?: string
) => {
  if (process.env.NODE_ENV !== 'production') {
    console.warn(
      `[shineout] '${prop}' in ${component} component is deprecated, please use '${newProp}' instead. ${extraMessage}`
    );
  }
};

const breakingChange = (message: string) => {
  if (process.env.NODE_ENV !== 'production') {
    console.warn(`[shineout] ${message}`);
  }
};

const devWarn = (message: string) => {
  if (process.env.NODE_ENV !== 'production') {
    console.warn(`[shineout] ${message}`);
  }
};

export const devUseWarning = {
  deprecated,
  warning: devWarn,
  breaking: breakingChange,
};
