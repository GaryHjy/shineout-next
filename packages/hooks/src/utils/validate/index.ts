import { ObjectType } from '../../common/type';
import { FormItemRule } from '../rule/rule.type';
import getRule from './get-rule';
import { flattenArray } from '../array';
import { wrapFormError } from './error';
export * from './error';

const handleProps = (props: ObjectType) => {
  return Object.keys(props).reduce((acc, key) => {
    const value = props[key];
    const type = typeof value;
    if (type === 'string' || type === 'number') {
      acc[key] = value;
    }
    return acc;
  }, {} as ObjectType);
};

export const validate = <T>(
  value: T | undefined,
  formData: ObjectType,
  rules: FormItemRule<T>,
  p: ObjectType,
) =>
  new Promise<true | Error>((resolve, reject) => {
    if (!rules.length) {
      resolve(true);
      return;
    }
    const props = handleProps(p);
    const [rule, ...restRule] = flattenArray(rules) as FormItemRule<T>;
    if (!rule) {
      validate(value, formData, restRule, props).then(resolve, reject);
      return;
    }
    const callback = (result: true | Error) => {
      if (result !== true) {
        reject(wrapFormError(result, props));
        return;
      }
      validate(value, formData, restRule, props).then(resolve, reject);
    };

    const fn = getRule(rule, props);
    const cb = fn(value, formData, callback, props);
    if (cb) {
      cb?.then(callback.bind(null, true)).catch((e: Error) => {
        reject(wrapFormError(e, props));
      });
    }
  });

export default validate;
