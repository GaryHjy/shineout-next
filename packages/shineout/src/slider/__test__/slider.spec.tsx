import { render, cleanup, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Slider from '..';
import { baseTest, createClassName, displayTest, snapshotTest } from '../../tests/utils';
import mountTest from '../../tests/mountTest';
import SliderBase from '../__example__/01-base';
import SliderRange from '../__example__/02-range';
import SliderScale from '../__example__/03-scale';
import SliderInput from '../__example__/04-input';
import SliderFormat from '../__example__/05-format';
import SliderStep from '../__example__/06-1-step';
import SliderStepZero from '../__example__/06-2-step-0';
import SliderHiderAuto from '../__example__/07-1-hide-auto';
import SliderHiderAll from '../__example__/07-2-hide-all';
import SliderDisabled from '../__example__/08-disabled';
import SliderVirtual from '../__example__/09-virtual';
import SliderIncrease from '../__example__/10-increase';
import SliderValueHover from '../__example__/11-value-hover';

const SO_PREFIX = 'slider';
const originClasses = ['wrapper'];
const originItemClasses = [''];
const { wrapper } = createClassName(SO_PREFIX, originClasses, originItemClasses);

afterEach(cleanup);
mountTest(<Slider />);

describe('Slider[Base]', () => {
  displayTest(Slider, 'ShineoutSlider');
  baseTest(Slider, wrapper);
  snapshotTest(<SliderBase />);
  snapshotTest(<SliderRange />, 'about range');
  snapshotTest(<SliderScale />, 'about scale');
  snapshotTest(<SliderInput />, 'about input');
  snapshotTest(<SliderFormat />, 'about format');
  snapshotTest(<SliderStep />, 'about step');
  snapshotTest(<SliderStepZero />, 'about step zero');
  snapshotTest(<SliderHiderAuto />, 'about hider auto');
  snapshotTest(<SliderHiderAll />, 'about hider all');
  snapshotTest(<SliderDisabled />, 'about disabled');
  snapshotTest(<SliderVirtual />, 'about virtual');
  snapshotTest(<SliderIncrease />, 'about increase');
  snapshotTest(<SliderValueHover />, 'about value hover');
  test('should render default', () => {
    render(<Slider />);
    screen.debug();
  });
});
