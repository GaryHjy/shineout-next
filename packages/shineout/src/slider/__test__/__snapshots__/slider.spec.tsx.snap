// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider[Base] should render correctly  1`] = `
<div
  class="cmzo79z"
>
  <div
    class="cj1pxn3"
  >
    <div
      class="ccmbr"
      style="left: 0%; right: 50%;"
    >
      <div
        class="c-7ssxpp c-uua2rc"
      />
      <div
        class="c-5umq26 cj2iwmt"
      >
        50
      </div>
    </div>
  </div>
  <div
    class="c54lued"
  >
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        0
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        100
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about disabled 1`] = `
<div
  class="cmzo79z c-trdv7s"
>
  <div
    class="cj1pxn3"
  >
    <div
      class="ccmbr"
      style="left: 25%; right: 25%;"
    >
      <div
        class="c-7ssxpp c-i77t4h"
      />
      <div
        class="cq16a97 cj2iwmt"
      >
        25
      </div>
      <div
        class="c-7ssxpp c-uua2rc"
      />
      <div
        class="c-5umq26 cj2iwmt"
      >
        75
      </div>
    </div>
  </div>
  <div
    class="c54lued"
  >
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        0
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        100
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about format 1`] = `
<div
  class="cmzo79z"
>
  <div
    class="cj1pxn3"
  >
    <div
      class="ccmbr"
      style="left: 6.111111111111112%; right: 60%;"
    >
      <div
        class="c-7ssxpp c-i77t4h"
      />
      <div
        class="cq16a97 cj2iwmt"
      >
        09:33
      </div>
      <div
        class="c-7ssxpp c-uua2rc"
      />
      <div
        class="c-5umq26 cj2iwmt"
      >
        12:36
      </div>
    </div>
  </div>
  <div
    class="c54lued"
  >
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        09:00
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        10:00
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        11:00
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        12:00
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        13:00
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        14:00
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        15:00
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        16:00
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        17:00
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        18:00
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about hider all 1`] = `
<div
  class="cmzo79z"
>
  <div
    class="cj1pxn3"
  >
    <div
      class="ccmbr"
      style="left: 0%; right: 96%;"
    >
      <div
        class="c-7ssxpp c-uua2rc"
      />
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about hider auto 1`] = `
<div
  class="cmzo79z c-agbuwj"
>
  <div
    class="cj1pxn3"
  >
    <div
      class="ccmbr"
      style="left: 0%; right: 75%;"
    >
      <div
        class="c-7ssxpp c-uua2rc"
      />
      <div
        class="c-5umq26 cj2iwmt"
      >
        4
      </div>
    </div>
  </div>
  <div
    class="c54lued"
  >
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        1
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        2
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        3
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        5
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        8
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        13
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        21
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        34
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        55
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        89
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        144
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about increase 1`] = `
<div>
  <div
    class="cmzo79z"
    style="margin-bottom: 24px;"
  >
    <div
      class="cj1pxn3"
    >
      <div
        class="ccmbr"
        style="left: 0%; right: 50%;"
      >
        <div
          class="c-7ssxpp c-uua2rc"
        />
        <div
          class="c-5umq26 cj2iwmt"
        >
          50
        </div>
      </div>
    </div>
    <div
      class="c54lued"
    >
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          0
        </div>
      </div>
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          100
        </div>
      </div>
    </div>
  </div>
  <div
    class="cmzo79z"
  >
    <div
      class="cj1pxn3"
    >
      <div
        class="ccmbr"
        style="left: 20%; right: 50%;"
      >
        <div
          class="c-7ssxpp c-i77t4h"
        />
        <div
          class="cq16a97 cj2iwmt"
        >
          20
        </div>
        <div
          class="c-7ssxpp c-uua2rc"
        />
        <div
          class="c-5umq26 cj2iwmt"
        >
          50
        </div>
      </div>
    </div>
    <div
      class="c54lued"
    >
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          0
        </div>
      </div>
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          100
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about input 1`] = `
<div
  style="display: flex; align-items: center;"
>
  <div
    class="cmzo79z"
    style="flex: 1; margin-inline-end: 28px;"
  >
    <div
      class="cj1pxn3"
    >
      <div
        class="ccmbr"
        style="left: 0%; right: 50%;"
      >
        <div
          class="c-7ssxpp c-uua2rc"
        />
        <div
          class="c-5umq26 cj2iwmt"
        >
          50
        </div>
      </div>
    </div>
    <div
      class="c54lued"
    >
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          0
        </div>
      </div>
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          100
        </div>
      </div>
    </div>
  </div>
  <div
    class="c62jszx c5g1ick"
    data-soui-type="input"
    style="width: 100px;"
  >
    <div
      class="cgc7yq7 c-dwl4jz cdip7we c-57d5o6"
    >
      <input
        class="cccsm1n"
        type="text"
        value="50"
      />
    </div>
    <div
      class="c-jaruk"
    >
      <span>
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.7071 12.7071C17.0676 12.3466 17.0953 11.7794 16.7903 11.3871L16.7071 11.2929L10.7071 5.29289C10.3166 4.90237 9.68342 4.90237 9.29289 5.29289C8.93241 5.65338 8.90468 6.22061 9.2097 6.6129L9.29289 6.70711L13.8779 11.2928C14.2683 11.6835 14.2683 12.3165 13.8779 12.7072L9.29289 17.2929C8.93241 17.6534 8.90468 18.2206 9.2097 18.6129L9.29289 18.7071C9.65338 19.0676 10.2206 19.0953 10.6129 18.7903L10.7071 18.7071L16.7071 12.7071Z"
          />
        </svg>
      </span>
      <span>
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.29289 12.7071C6.93241 12.3466 6.90468 11.7794 7.2097 11.3871L7.29289 11.2929L13.2929 5.29289C13.6834 4.90237 14.3166 4.90237 14.7071 5.29289C15.0676 5.65338 15.0953 6.22061 14.7903 6.6129L14.7071 6.70711L10.1221 11.2928C9.7317 11.6835 9.7317 12.3165 10.1221 12.7072L14.7071 17.2929C15.0676 17.6534 15.0953 18.2206 14.7903 18.6129L14.7071 18.7071C14.3466 19.0676 13.7794 19.0953 13.3871 18.7903L13.2929 18.7071L7.29289 12.7071Z"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about range 1`] = `
<div
  class="cmzo79z"
>
  <div
    class="cj1pxn3"
  >
    <div
      class="ccmbr"
      style="left: 25%; right: 25%;"
    >
      <div
        class="c-7ssxpp c-i77t4h"
      />
      <div
        class="cq16a97 cj2iwmt"
      >
        25
      </div>
      <div
        class="c-7ssxpp c-uua2rc"
      />
      <div
        class="c-5umq26 cj2iwmt"
      >
        75
      </div>
    </div>
  </div>
  <div
    class="c54lued"
  >
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        0
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        100
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about scale 1`] = `
<div
  class="cmzo79z"
>
  <div
    class="cj1pxn3"
  >
    <div
      class="ccmbr"
      style="left: 0%; right: 100%;"
    >
      <div
        class="c-7ssxpp c-uua2rc"
      />
      <div
        class="c-5umq26 cj2iwmt"
      >
        0
      </div>
    </div>
  </div>
  <div
    class="c54lued"
  >
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        0
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        10
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        100
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        250
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        500
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        1000
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about step 1`] = `
<div
  class="cmzo79z"
>
  <div
    class="cj1pxn3"
  >
    <div
      class="ccmbr"
      style="left: 5%; right: 75%;"
    >
      <div
        class="c-7ssxpp c-i77t4h"
      />
      <div
        class="cq16a97 cj2iwmt"
      >
        0.05
      </div>
      <div
        class="c-7ssxpp c-uua2rc"
      />
      <div
        class="c-5umq26 cj2iwmt"
      >
        0.25
      </div>
    </div>
  </div>
  <div
    class="c54lued"
  >
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        0
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        1
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about step zero 1`] = `
<div
  class="cmzo79z"
>
  <div
    class="cj1pxn3"
  >
    <div
      class="ccmbr"
      style="left: 0%; right: 100%;"
    >
      <div
        class="c-7ssxpp c-uua2rc"
      />
      <div
        class="c-5umq26 cj2iwmt"
      >
        0.8
      </div>
    </div>
  </div>
  <div
    class="c54lued"
  >
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        0.8
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        1
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        1.2
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        1.4
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        1.7
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        2
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        2.4
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        2.8
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        3.3
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        4
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        4.8
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        5.6
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        6.7
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        8
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        9.5
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        11
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        13
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        16
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about value hover 1`] = `
<div
  class="cmzo79z"
>
  <div
    class="cj1pxn3"
  >
    <div
      class="ccmbr"
      style="left: 0%; right: 50%;"
    >
      <div
        class="c-7ssxpp c-uua2rc c-idcbbr"
      />
      <div
        class="c-5umq26 cj2iwmt celru6f"
      >
        50
      </div>
    </div>
  </div>
  <div
    class="c54lued"
  >
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        0
      </div>
    </div>
    <div
      class="cj0wkge"
    >
      <div
        class="cix0qwo"
      >
        100
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider[Base] should render correctly about virtual 1`] = `
<div>
  <div
    class="cmzo79z c3z3y5u"
    style="height: 200px;"
  >
    <div
      class="cj1pxn3"
    >
      <div
        class="ccmbr"
        style="bottom: 0%; top: 50%;"
      >
        <div
          class="c-7ssxpp c-uua2rc"
        />
        <div
          class="c-5umq26 cj2iwmt"
        >
          50
        </div>
      </div>
    </div>
    <div
      class="c54lued"
    >
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          0
        </div>
      </div>
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          100
        </div>
      </div>
    </div>
  </div>
  <div
    class="cmzo79z c3z3y5u"
    style="height: 200px;"
  >
    <div
      class="cj1pxn3"
    >
      <div
        class="ccmbr"
        style="bottom: 12%; top: 30.000000000000004%;"
      >
        <div
          class="c-7ssxpp c-i77t4h"
        />
        <div
          class="cq16a97 cj2iwmt"
        >
          12
        </div>
        <div
          class="c-7ssxpp c-uua2rc"
        />
        <div
          class="c-5umq26 cj2iwmt"
        >
          70
        </div>
      </div>
    </div>
    <div
      class="c54lued"
    >
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          0
        </div>
      </div>
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          100
        </div>
      </div>
    </div>
  </div>
  <div
    class="cmzo79z c3z3y5u"
    style="height: 200px;"
  >
    <div
      class="cj1pxn3"
    >
      <div
        class="ccmbr"
        style="bottom: 0%; top: 77.5%;"
      >
        <div
          class="c-7ssxpp c-uua2rc"
        />
      </div>
    </div>
    <div
      class="c54lued"
    >
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          0℃
        </div>
      </div>
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          20℃
        </div>
      </div>
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          40℃
        </div>
      </div>
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          60℃
        </div>
      </div>
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          100℃
        </div>
      </div>
    </div>
  </div>
  <div
    class="cmzo79z c-agbuwj c3z3y5u"
    style="height: 200px;"
  >
    <div
      class="cj1pxn3"
    >
      <div
        class="ccmbr"
        style="bottom: 12%; top: 30.000000000000004%;"
      >
        <div
          class="c-7ssxpp c-i77t4h"
        />
        <div
          class="cq16a97 cj2iwmt"
        >
          12
        </div>
        <div
          class="c-7ssxpp c-uua2rc"
        />
        <div
          class="c-5umq26 cj2iwmt"
        >
          70
        </div>
      </div>
    </div>
    <div
      class="c54lued"
    >
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          0
        </div>
      </div>
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          100
        </div>
      </div>
    </div>
  </div>
  <div
    class="cmzo79z c-trdv7s c3z3y5u"
    style="height: 200px;"
  >
    <div
      class="cj1pxn3"
    >
      <div
        class="ccmbr"
        style="bottom: 12%; top: 30.000000000000004%;"
      >
        <div
          class="c-7ssxpp c-i77t4h"
        />
        <div
          class="cq16a97 cj2iwmt"
        >
          12
        </div>
        <div
          class="c-7ssxpp c-uua2rc"
        />
        <div
          class="c-5umq26 cj2iwmt"
        >
          70
        </div>
      </div>
    </div>
    <div
      class="c54lued"
    >
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          0
        </div>
      </div>
      <div
        class="cj0wkge"
      >
        <div
          class="cix0qwo"
        >
          100
        </div>
      </div>
    </div>
  </div>
</div>
`;
